storage/repair/types.go

```diff
@@ -0,0 +1,120 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package repair
+
+import (
+	"time"
+
+	"github.com/m3db/m3db/client"
+	"github.com/m3db/m3db/storage/block"
+	"github.com/m3db/m3db/topology"
+)
+
+// HostBlockMetadata contains a host along with block metadata from that host
+type HostBlockMetadata struct {
+	Host     topology.Host
+	Size     int64
+	Checksum *uint32
+}
+
+// ReplicaBlockMetadata captures the block metadata from hosts in a shard replica set
+type ReplicaBlockMetadata interface {
+	// Start is the start time of a block
+	Start() time.Time
+
+	// Metadata returns the metadata from all hosts
+	Metadata() []HostBlockMetadata
+
+	// Add adds a metadata from a host
+	Add(metadata HostBlockMetadata)
+}
+
+// ReplicaBlocksMetadata captures the blocks metadata from hosts in a shard replica set
+type ReplicaBlocksMetadata interface {
+	// Blocks returns the blocks metadata
+	Blocks() map[time.Time]ReplicaBlockMetadata
+
+	// Add adds a block metadata
+	Add(block ReplicaBlockMetadata)
+
+	// GetOrAdd returns the blocks metadata for a start time, creating one if it doesn't exist
+	GetOrAdd(start time.Time) ReplicaBlockMetadata
+}
+
+// ReplicaSeriesMetadata captures the metadata for a list of series from hosts in a shard replica set
+type ReplicaSeriesMetadata interface {
+	// Series returns the series metadata
+	Series() map[string]ReplicaBlocksMetadata
+
+	// GetOrAdd returns the series metadata for an id, creating one if it doesn't exist
+	GetOrAdd(id string) ReplicaBlocksMetadata
+}
+
+// ReplicaMetadataComparer compares metadata from hosts in a replica set
+type ReplicaMetadataComparer interface {
+	// AddLocalMetadata adds metadata from local host
+	AddLocalMetadata(origin topology.Host, localIter block.FilteredBlocksMetadataIter)
+
+	// AddPeerMetadata adds metadata from peers
+	AddPeerMetadata(peerIter client.PeerBlocksMetadataIter) error
+
+	// Compare returns the metadata differences between local host and peers
+	Compare() MetadataComparisonResult
+}
+
+// MetadataComparisonResult captures metadata comparison results
+type MetadataComparisonResult struct {
+	// NumSeries returns the total number of series
+	NumSeries int64
+
+	// NumBlocks returns the total number of blocks
+	NumBlocks int64
+
+	// SizeResult returns the size differences
+	SizeDifferences ReplicaSeriesMetadata
+
+	// ChecksumDifferences returns the checksum differences
+	ChecksumDifferences ReplicaSeriesMetadata
+}
+
+// Options are the repair options
+type Options interface {
+	// SetNewAdminSessionFn sets the new admin session function
+	SetNewAdminSessionFn(value client.NewAdminSessionFn) Options
+
+	// NewAdminSessionFn returns the new admin session function
+	NewAdminSessionFn() client.NewAdminSessionFn
```

Comment:
That's fine, opening a session before opening the database is ok.  That's why I implemented the `ConnectConsistencyLevelAny` so that you can attempt to connect to all, then to majority, then to one and then to none.